keyboards/darkproject/kd83a_ltd/matrix.c

```diff
@@ -0,0 +1,87 @@
+/*
```

Comment:
> This PR features custom matrix scanning code with adaptive scan rate and anti-ghosting.

Due to the maintenance cost involved, this code should be removed and the core implementation used instead.

Reply:
This code is simple and efficient, it uses basic C constructions and PAL to improve it's reliability. It also takes into account the individual features of the structure and layout of the keyboard. The main algorithm of key scanning is robust and this is probably the only way to register keypresses for this hardware configuration.

Reply:
The issue isnt that myself or other maintainers would have failed to understand the code. The issue is that we do not want to support it. Anything that is effectively duplicating core functionality is increasing the ongoing maintenance burden to QMK and its maintainers. Especially when that functionality is perfectly acceptable for thousands of keyboards and multiples more users.

Going forward we are focusing on making keyboards follow ["data driven"](https://docs.qmk.fm/#/data_driven_config). This is to increase maintainability. custom matrix implementations for speed would be the opposite direction to where QMK is heading.

Reply:
My take was not about understanding but about its simplicity from PAL base side which should not require that much maintenance if at all.
Can I apply or add custom matrix implementation only for certain user keymaps then and maintain it later by myself? VIA and default keymaps will use core matrix scanning routine.

Reply:
Longer term, there had been some discussion to code-gen more optimised matrix scanning code. We also had some users discussing enhancing the core routines with the pin asserting wait logic. 

However in the short term, the code should just not exist in this repo. You are welcome to host "optimised" implementations elsewhere?

Reply:
Sorry for late reply. I was investigating problem with ghosting and phantom keypresses on some boards. Now adding custom matrix is mandatory because it resolves it.
I will support this code and keep it up to QMK standarts.
This applies to both KD83a LTD and KD87b LTD.

Reply:
From my perspective, still needs to use core and wouldn't be accepted till the code is removed.

Reply:
The keyboards have issues with core implementation. In this case users can experience some random problems. I have some user reports. Custom implementation does not have these issues and has filtering algorithm.

Reply:
Given the insistence to use the custom implementation over core before the discovered issues, I have doubts it has been given a proper test or that the existing options available exercised.

Reply:
The issue is quite random and hard to represent and has been discovered recently. My keyboard does not have this issue but several other boards have. With core implementation I was unable to found really stable configuration.
I added pretty strict pin debounce algorithm which resolved the issue and all affected boards.
All code in this PR is quite excessively tested within several copies of keyboard and proven to work.
From my perspective the best is adding custom matrix code. In this scenario I can be sure that users which will compile this firmware won't experience these random issues. Of course if I will recieve user reports with another problems I will investigate and update firmware source to provide fixes.